src/vs/workbench/services/positronDataExplorer/common/tableDataCache.ts

```diff
@@ -233,64 +229,112 @@ export class TableDataCache extends Disposable {
-				let longestDataCell = 0;
-				for (const dataCell of dataColumn.values()) {
-					if (dataCell.formatted.length > longestDataCell) {
-						longestDataCell = dataCell.formatted.length;
-					}
-				}
+	async calculateColumnLayoutEntries(
+		minimumColumnWidth: number,
+		maximumColumnWidth: number
+	): Promise<number[] | undefined> {
+		// If the width calculators are not set, return undefined.
+		if (!this._widthCalculators) {
+			return undefined;
+		}
+
+		// Get the number of columns. If there are no columns, or too many columns, return
+		// undefined.
+		const tableState = await this._dataExplorerClientInstance.getBackendState();
+		const { num_columns: numColumns } = tableState.table_shape;
+		if (!numColumns || numColumns > MAX_AUTO_SIZE_COLUMNS) {
+			return undefined;
+		}
 
-				// If the longest data cell is non-zero, calculate and cache its width.
-				this._columnValueWidthCache.set(
-					columnIndex,
-					this._columnValueWidthCalculator(longestDataCell)
+		// Page in schema and data for column width calculations.
+		let columnIndex = 0;
+		const columnWidths: number[] = [];
+		while (columnIndex < numColumns) {
+			// Calculate the page size of the page of schema and the page of data to load.
+			const pageSize = Math.min(AUTO_SIZE_COLUMNS_PAGE_SIZE, numColumns - columnIndex);
+
+			// Load the page of schema.
+			const tableSchema = await this._dataExplorerClientInstance.getSchema(
+				Array.from({ length: pageSize }, (_, i) => columnIndex + i)
+			);
+
+			// Calculate the column header widths for the page of schema that was returned.
+			for (let i = 0; i < tableSchema.columns.length; i++) {
+				const columnSchema = tableSchema.columns[i];
+				columnWidths[columnIndex + i] = pinToRange(
+					this._widthCalculators.columnHeaderWidthCalculator(
+						columnSchema.column_name,
+						columnSchema.type_name
+					),
+					minimumColumnWidth,
+					maximumColumnWidth
 				);
 			}
+
+			// Load the page of data.
+			const tableData = await this._dataExplorerClientInstance.getDataValues(
+				Array.from({ length: pageSize }, (_, i): ColumnSelection => ({
+					column_index: columnIndex + i,
+					spec: {
+						first_index: 0,
+						last_index: 9
+					}
+				}))
+			);
+
+			// Calculate the column value widths for the page of data that was returned.
+			for (let column = 0; column < tableData.columns.length; column++) {
+				for (let row = 0; row < tableData.columns[column].length; row++) {
+					// Get the cell value.
+					const value = tableData.columns[column][row];
+
+					// Convert the cell value into a data cell.
+					let dataCell: DataCell;
+					if (typeof value === 'number') {
+						dataCell = decodeSpecialValue(value);
+					} else {
+						dataCell = {
+							formatted: value,
+							kind: DataCellKind.NON_NULL
+						};
+					}
+
+					// Calculate the column value width.
+					const columnValueWidth = pinToRange(
+						this._widthCalculators.columnValueWidthCalculator(
```

Comment:
How expensive is it to run this calculator on every value in the page of data? (Wondering if e.g. we could run it on only the longest strings for character columns after some heuristics, etc.)